### PR TITLE
DateProcessor Locale

### PR DESCRIPTION
**Pull request by sneivandt**
Set the language on the Locale.Builder used to construct the DateProcessor. This change results in the locale that is generated being equal to one of standard locales such as Locale.SPANISH given the localeString "SPANISH".

This change is for issue #25513.

### CODE REVIEW

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateProcessor.java [hunk 1]
```diff
@@ -121,7 +122,16 @@ public DateProcessor create(Map<String, Processor.Factory> registry, String proc
             Locale locale = Locale.ENGLISH;
             if (localeString != null) {
                 try {
-                    locale = (new Locale.Builder()).setLanguageTag(localeString).build();
```
Comment (jpountz) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateProcessor.java]: would it work to use `LocalUtils.parse(localeString)` like we do for the date field?

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateFormat.java [hunk 2]
```diff
@@ -67,7 +67,7 @@ private long parseMillis(String date) {
         @Override
         Function<String, DateTime> getFunction(String format, DateTimeZone timezone, Locale locale) {
             DateTimeFormatter parser = DateTimeFormat.forPattern(format).withZone(timezone).withLocale(locale);
-            return text -> parser.withDefaultYear((new DateTime(DateTimeZone.UTC)).getYear()).parseDateTime(text);
+            return text -> parser.withDefaultYear((new DateTime(DateTimeZone.UTC)).getYear()).parseDateTime(text.toLowerCase());
```
Comment (jpountz) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateFormat.java]: this feels wrong to me, why did you do that change?

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateProcessor.java (hunk 1 above)
Comment (sneivandt) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateProcessor.java]: This does not have the desired result. This snippet evaluates as false.

`LocaleUtils.parse("FRENCH").getLanguage().equals(Locale.FRENCH.getLanguage())`

The locale from LocaleUtils has the language "french" while Locale.FRENCH has the language "fr". This seems to be the key difference that results in the JodaTime parser not correctly looking for words in the correct language.

Perhaps it makes sense to move the logic that I have written here to convert from the display name for a language to the ISO code.

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateFormat.java (hunk 2 above)
Comment (sneivandt) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateFormat.java]: This is an attempt to make the date parsing case insensitive. If you remove this change then you will be able to parse "mai 2017" but not "Mai 2017" for example.

Comment (jpountz) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateFormat.java]: It looks a bit wrong to me. If you would like to suggest this change, I'd like it to be a separate PR.

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateProcessor.java (hunk 1 above)
Comment (jpountz) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateProcessor.java]: Why do you want `FRENCH` to work? The actual locale name is `fr`, not `FRENCH`?

**Review by jpountz — CHANGES_REQUESTED** [review]
I don't think we should try to understand the names of Java constants as locale names. However I realized that we parse locales differently in the date processor and in date fields, which I think we should fix.

**Comment by jpountz** [comment]
I'm wondering whether a better fix would be to fail if the language code has less than 2 or more than 3 letters? This way it would be less trappy as `FRENCH` would fail immediately instead of being silently ignored?

**Comment by sneivandt** [comment]
@jpountz I updated the change so that now it simply validates that the language tag matches an available locale. In this state it will not accept `FRENCH` as a valid locale but would accept values such as `fr` or `fr-CA`.

**Comment by jpountz** [comment]
I have a limited knowledge of locales and language tags, but such a change makes me a bit worried that some valid tags may be rejected eg. because they are aliases and `Locale.getAvailableLocales()` will only give you the canonical ones.

**Comment by sneivandt** [comment]
I agree with your concerns about filtering out potentially valid locales. I am not certain if there is a legitemate use case that would be impacted by validating against locales in `Locale.getAvailableLocales()`. I have updated this PR so that it just does does a simple check based on the languageString. This is a safer check and will prevent confusing errors as a result of using a value like `FRENCH`. You can of course still construct something with a value like `AA-AA` that the Locale.Builder will still accept when constructing the locale.

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateProcessor.java [hunk 3]
```diff
@@ -125,6 +125,11 @@ public DateProcessor create(Map<String, Processor.Factory> registry, String proc
                 } catch (IllformedLocaleException e) {
                     throw new IllegalArgumentException("Invalid language tag specified: " + localeString);
                 }
+                for (String part : localeString.split("-")) {
+                    if (part.length() < 2 || part.length() > 3) {
+                        throw new IllegalArgumentException("Invalid language tag specified: " + localeString);
+                    }
+                }
```
Comment (jpountz) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateProcessor.java]: A single value should still work?

Maybe we could use the validation logic from https://stackoverflow.com/questions/3684747/how-to-validate-a-locale-in-java in order to make sure that the provided locale is recognized by the JVM?

**Review by jpountz — CHANGES_REQUESTED** [review]

**Review by jpountz — APPROVED** [review]

**Comment by jpountz** [comment]
Thanks @sneivandt !

**Pull request merged**